### PR TITLE
Self-supevised consistency loss for MuZero

### PR DESCRIPTION
**Pull request by emailweixu**
As reported in [Ye et. al. Mastering Atari Games with Limited Data](https://arxiv.org/abs/2111.00210), the consistency loss for latent representation is useful. This PR implements it.

Tested on pendulum, does not show significant difference, possibly because pendulum is too simple.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -2105,6 +2105,9 @@ def __init__(self, shape):
             shape (tuple): desired shape not including the batch dimension.
```
Comment (breakds) [inline alf/layers.py]: The type hint `(tuple)` here is not precise after the change.

File: alf/algorithms/mcts_models.py [hunk 2]
```diff
@@ -75,36 +75,69 @@
...
+                ``representation_net(observation)``.
+            dynamics_net (Network): the network for generating the next latent
+                representation given the current latent representation and action.
+                It is called as ``dynamics_net((current_latent_representation, action))``
             train_reward_function (bool): whether to predict reward
             train_game_over_function (bool): whether to predict game over
+            train_repr_prediction (bool): whether to train to predict future
+                latent representation.
+            predict_reward_sum (bool): If True, the loss for reward is the mean
+                square error between the sum of predicted reward and the sum of
```
Comment (breakds) [inline alf/algorithms/mcts_models.py]: Might be good to explicitly state how (across what) do we compute the `sum` of reward to help the user understand.

File: alf/algorithms/mcts_models.py [hunk 3]
```diff
@@ -412,8 +518,29 @@ def __init__(self,
             alpha_adjust_rate (float): the speed to adjust alpha
             train_reward_function (bool): whether to predict reward
             train_game_over_function (bool): whether to predict game over
+            train_repr_prediction (bool): whether to train to predict future
+                latent representation.  This implements the self-supervised
+                consistency loss described in `Ye et. al. Mastering Atari Games
+                with Limited Data <https://arxiv.org/abs/2111.00210>`_. The loss
+                is ``-cosine(prediction_net(projection_net(x), projection_net(y))``,
```
Comment (breakds) [inline alf/algorithms/mcts_models.py]: Seems that there is a missing `)` right after `projection_net(x)`.

**Review by breakds — DISMISSED** [review]
LGTM. Thanks for making it easier to read and follow!

File: alf/algorithms/muzero_algorithm.py [hunk 4]
```diff
@@ -343,6 +363,24 @@ def preprocess_experience(self, root_inputs: TimeStep,
...
+                    observation, step_type = replay_buffer.get_field(
+                        ('observation', 'step_type'), env_ids, positions)
+                    exp = alf.data_structures.make_experience(
+                        root_inputs, AlgStep(), state=())
+                    exp = exp._replace(
+                        step_type=step_type,
+                        observation=observation,
+                        batch_info=batch_info,
+                        replay_buffer=replay_buffer)
+                    exp = self._data_transformer.transform_experience(exp)
```
Comment (Haichao-Zhang) [inline alf/algorithms/muzero_algorithm.py:381]: I might mis-understood this part. 
But I'm curious why do we need to ``transform_experience`` when ``type(self._data_transformer)`` is  an ``IdentityDataTransformer``? 
We should ``transform_experience`` when it is not an ``IdentityDataTransformer``?

Comment (emailweixu) [reply alf/algorithms/muzero_algorithm.py:381]: Thanks for spotting it. This is introduced when I cleaned up for PR.

**Review by Haichao-Zhang — APPROVED** [review]

**Review by breakds — APPROVED** [review]

**Pull request merged**